internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -1058,3 +1072,77 @@ func (a *Adapter) findSnapshotWithOpenedPR(snapshot *applicationapiv1alpha1.Snap
 	}
 	return nil, nil
 }
+
+// checkOldSnapshotsPipelineRun sorts all snapshots for application and cancells all running interationTest pipelineruns within application
+// removes finalizer before the pipelinerun is set as cancelled run finally to be gracefuly cancelled
+func (a *Adapter) checkAndCancelOldSnapshotsPipelineRun(application *applicationapiv1alpha1.Application, snapshot *applicationapiv1alpha1.Snapshot) error {
+
+	snapshots, err := a.loader.GetAllSnapshotsForPR(a.context, a.client, application, snapshot.GetLabels()[gitops.PipelineAsCodePullRequestAnnotation], snapshot.GetLabels()[gitops.PipelineAsCodeRepoURLAnnotation])
+	if err != nil {
+		a.logger.Error(err, "Failed to fetch Snapshots for the application",
+			"application.Name:", application.Name)
+		return err
+	}
+	sortedSnapshots := gitops.SortSnapshots(*snapshots)
+	if len(sortedSnapshots) < 1 {
```

Comment:
Just realized that maybe this should be checking `if len(sortedSnapshots) < 2`.
If we have just one Snapshot (let's say the current one) in the list, and the `for` loop below starts looping from the index 1, that could cause an index out of range error.

Reply:
Never mind, I just tested it out and the initial condition in the for loop has been well set and catches it correctly.

Reply:
That is a good point, more than one snapshots is required for this function to work correctly.